File: Programmers/단어변환.py
result = list()


def solution(begin, target, words):
    global result
    answer = 0
    if target not in words:
        return 0
    elif target == begin:
        return 0

    visited = [False] * len(words)
    dfs(begin, target, words, visited, answer)
    return min(result)


def dfs(begin, target, words, visited, answer):
    global result
    if begin == target:
        return result.append(answer)

    change = list()

    for word in words:
        count = 0
        for i in range(len(word)):
            if begin[i] != word[i]:
                count += 1
        if count == 1:
            change.append(word)

    for word in change:
        if not visited[words.index(word)]:
            visited[words.index(word)] = True
            dfs(word, target, words, visited, answer + 1)
            visited[words.index(word)] = False
solution("hit", "cog", a)

File: Java/src/com/company/level3/단어변환.java
package com.company.level3;

public class 단어변환 {
    int globalAnswer = 1000;
    public int solution(String begin, String target, String[] words) {
        int answer = 0;

        dfs(begin, target, words, new boolean[words.length], 0);
        answer = globalAnswer;
        if(answer == 1000) answer = 0;
        return answer;
    }

    public void dfs(String begin, String target, String[] words, boolean[] visited, int depth){
        if(begin.equals(target)){
            globalAnswer = Math.min(globalAnswer, depth);
            return;
        }
        if(depth == words.length){
            return;
        }

        for(int i = 0; i < words.length; i++){
            if(visited[i]) continue;

            String s = words[i];
            String temp = begin;
            int different = 0;
            for(int j = 0; j < s.length(); j++){
                if(begin.charAt(j) != s.charAt(j)){
                    different += 1;
                }
            }

            if(different != 1) continue;

            begin = s;
            visited[i] = true;
            dfs(begin, target, words, visited, depth+1);
            visited[i] = false;
            begin = temp;
        }
    }
}
